### DISCUSSION TITLE
How to handle complex deployments where branching is not enough?

### DISCUSSION THREAD

**Discussion by geoHeil** [dagster-io/dagster#23227] [Q&A]
Currently, dagster branch deployments do not allow the execution of sensors or freshness policies.
This means - branch deployments are great - for unit testing the assets in isolation.
Unit testing the interaction of the freshness policies will improve soon https://github.com/dagster-io/dagster/discussions/22811

However, sensors - and to some extent the real experience of how the policies interact with each other and the world- are still missing.

How do you solve this challenge?

Do you use a git main branch for a prod deploy, feature branches to test these in isolation and an additional staging branch that deploys to a separate dagster (full) deployment target to test sensors, asset interaction and freshness policies?

**Comment by zyd14** [comment]
We do github releases for prod deployment, pushes to main get deployed to dev. Sensors and schedules and that stuff mostly get tested in dev. We have an additional staging deployment for running automated full-scale end to end test pipelines as part of the release CI process. We don't really test sensors in staging as staging is mostly just set up for automated integration tests during release process.

**Reply by geoHeil** [reply]
Why did you decide for this structure? To be more safe? I.e. no automatic deploy? Or in order to better test the sensors?

**Reply by geoHeil** [reply]
Do you think that with the new upcoming and their better testability https://github.com/dagster-io/dagster/discussions/22811 there will be less need for such a (more complex) setup?

**Reply by saksham1525** [reply]
Hi again @zyd14, I’m facing a major issue due to sensors not being available on branch deployments. My pipeline relies on a sensor right at the ingestion step — it continuously polls for SQS queue messages, extracts relevant details, creates partitions based on that information, and then outputs the information extracted from SQS records to a downstream asset.
Since branch deployments don’t support sensors, I’m unable to ingest data within the branch pipeline. This makes branch testing effectively redundant for my use case. To test any changes, I currently have to push them to the gitlab's main branch (using dagster's dev env for testing), which clutters the repo and becomes difficult to manage when multiple people are developing and testing in parallel.

Could you please clarify why this restriction was introduced? Was it driven by cost considerations or some other limitation? Is there a solution for my use case?

**Reply by geoHeil** [reply]
solution: do a full blown pre-production environment where the sensors are active

**Reply by zyd14** [reply]
@saksham1525 If you're not able to create a new pre-production deployment (not on an Enterprise plan) you could consider the [sensor testing feature](https://docs.dagster.io/guides/automate/sensors/testing-sensors) Dagster+ has. 

I don't know exactly why sensors are not able to be ran on branch deployments, I'm not part of the Dagster team so I can only speculate. It may have something to do with the fact that branch deployments have one agent serving many different branch deployments - perhaps there is something about the agent architecture that makes it difficult to handle schedules and sensors for many different branch deployment code locations with a single agent.

edit - you could also do a OSS deployment to test your sensors with
